File: haskell/2015/Day3.hs
module Day3 where

import Data.List
import qualified Data.Map as M

move :: (Int, Int) -> Char -> (Int, Int)
move (x, y) '^' = (x, y + 1)
move (x, y) '>' = (x + 1, y)
move (x, y) 'v' = (x, y - 1)
move (x, y) '<' = (x - 1, y)
move p _ = p

main' :: IO ()
main' = do
  contents <- readFile "2015/inputs/Day3/input.txt"
  let contents' = scanl move (0, 0) contents
  -- part 1
  let results = foldr (\x -> M.insertWith (+) x 1) (M.singleton (0, 0) 1) contents'
  print $ M.size results
  -- part 2
  let a = zip [0 ..] contents
  let (santa, robot) = partition (\(i, _) -> even i) a
  let santa' = scanl move (0, 0) $ map snd santa
  let robot' = scanl move (0, 0) $ map snd robot
  let results2 = foldr (\x -> M.insertWith (+) x 1) (M.singleton (0, 0) 2) santa'
  let results2' = foldr (\x -> M.insertWith (+) x 1) results2 robot'
  print $ M.size results2'

File: elixir/2015/lib/day3.ex
defmodule Day3 do
  require Integer

  @spec move(char(), {integer(), integer()}) :: {integer(), integer()}
  def move(c, {x, y}) do
    case c do
      ?^ -> {x, y + 1}
      ?> -> {x + 1, y}
      ?v -> {x, y - 1}
      ?< -> {x - 1, y}
      _ -> {x, y}
    end
  end

  @spec main() :: any()
  def main() do
    contents =
      File.read!("../../inputs/2015/Day3/input.txt")

    parsed =
      to_charlist(contents)
      |> Enum.scan({0, 0}, &Day3.move(&1, &2))

    parsed = [{0, 0}] ++ parsed

    # part 1
    r1 = Enum.group_by(parsed, & &1) |> Map.keys() |> length()
    IO.inspect(r1)

    # part 2
    {santa, robot} =
      contents
      |> to_charlist()
      |> Enum.with_index()
      |> Enum.split_with(fn {_, i} -> Integer.is_even(i) end)

    santa =
      ([{0, 0}] ++ santa)
      |> Enum.map(&elem(&1, 0))
      |> Enum.scan({0, 0}, &Day3.move(&1, &2))

    robot =
      ([{0, 0}] ++ robot)
      |> Enum.map(&elem(&1, 0))
      |> Enum.scan({0, 0}, &Day3.move(&1, &2))

    r2 = Enum.group_by(santa ++ robot, & &1) |> Map.keys() |> length()
    IO.inspect(r2)
  end
end
